src/rtl/xdpworkqueue.c

```diff
@@ -288,7 +295,8 @@ XdpIoWorkItemRoutine(
         XdpWorkQueueReleaseLock(WorkQueue, OldIrql);
         KeSetEvent(WorkQueue->ShutdownEvent, 0, FALSE);
     }
-    XdpDereferenceWorkQueue(WorkQueue);
 
     TraceExitSuccess(TRACE_RTL);
+
+    XdpDereferenceWorkQueue(WorkQueue);
```

Comment:
Why trace exit _before_ dereferencing?

Reply:
This is a bit tricky, but the dereference protects the ETL tracing session.

Reply:
Ok. Maybe a comment is worth adding here, because I could totally see someone else come in and say "Oh, this looks backwards, let's swap these calls."

Reply:
Yeah, added. I left out the comment originally because nothing really bad would happen if we reverted to the more natural ordering. It would just make debugging this scenario slightly more unpredictable. It's still non-deterministic even with all these logs and refcounts.